<a href="https://colab.research.google.com/github/IrynaSukhoverkhova/my-first-work-place/blob/main/hw_10_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd
df = pd.read_csv('supermarket_sales.csv')

1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [4]:
branch_profit = df.groupby('Branch')['gross income'].sum()
print("Сумарний прибуток за філіями:\n", branch_profit)

Сумарний прибуток за філіями:
 Branch
A    5057.1605
B    5057.0320
C    5265.1765
Name: gross income, dtype: float64


In [6]:
best_branch = branch_profit.idxmax()
print(f"\nНайприбутковіший філіал: {best_branch}")


Найприбутковіший філіал: C


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [7]:
best_city = df[df['Branch'] == best_branch]['City'].iloc[0]
print(f"Філіал {best_branch} знаходиться у місті {best_city} ")

Філіал C знаходиться у місті Naypyitaw 


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [8]:
invoices_by_category = df.pivot_table(
    index='Branch',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

print(invoices_by_category)

Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Branch                                                                          
A                                 60                   51                  58   
B                                 55                   62                  50   
C                                 55                   65                  66   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Branch                                                                  
A                            47                  65                 59  
B                            53                  50                 62  
C                            52                  45                 45  


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [9]:
best_branch_electronics = invoices_by_category['Electronic accessories'].idxmax()
print(f"Найбільше інвойсів із покупкою товарів категорії 'Electronic accessories' у філії: {best_branch_electronics} ")

Найбільше інвойсів із покупкою товарів категорії 'Electronic accessories' у філії: A 


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


Питання 1. У який час доби (ранок, день, вечір) покупці найчастіше купують різні категорії товарів?

In [10]:
df['Hour'] = pd.to_datetime(df['Time']).dt.hour

def time_of_day(hour):
    if 9 <= hour < 12:
        return 'Ранок (9-12)'
    elif 12 <= hour < 17:
        return 'День (12-17)'
    else:
        return 'Вечір (17+)'

df['Time_Period'] = df['Hour'].apply(time_of_day)


sales_by_time = df.pivot_table(
    index='Time_Period',
    columns='Product line',
    values='Invoice ID',
    aggfunc='count'
)

print("Розподіл покупок за часом доби:")
print(sales_by_time)

Розподіл покупок за часом доби:
Product line  Electronic accessories  Fashion accessories  Food and beverages  \
Time_Period                                                                     
Вечір (17+)                       66                   67                  69   
День (12-17)                      69                   83                  74   
Ранок (9-12)                      35                   28                  31   

Product line  Health and beauty  Home and lifestyle  Sports and travel  
Time_Period                                                             
Вечір (17+)                  53                  51                 49  
День (12-17)                 72                  70                 86  
Ранок (9-12)                 27                  39                 31  


/tmp/ipykernel_1241/2483734876.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['Hour'] = pd.to_datetime(df['Time']).dt.hour


Для бізнесу:
9:00–12:00 - ранок є найменш прибутковим періодом для всіх категорій. Найнижчі показники зафіксовано у Health and beauty (27) та Fashion accessories (28).Відносно краще вранці продаються Home and lifestyle (39) та Electronic accessories (35).

12:00–17:00 - це найактивніший період для абсолютно всіх категорій товарів.Найвищий попит у денний час спостерігається в категоріях Sports and travel (86) та Fashion accessories (83).

Після 17:00 - ввечері активність залишається високою, особливо для Food and beverages (69), Fashion accessories (67) та Electronic accessories (66).Попит на Sports and travel суттєво падає порівняно з днем (з 86 до 49).

У ранкові години (9:00–12:00) можна зменшити кількість працівників у залі та на касах, залучаючи їх до викладки товару, інвентаризації або підготовки магазину.

Максимальну кількість касирів та консультантів варто виводити на зміну в проміжку з 12:00 до 17:00, щоб уникнути черг та втрати клієнтів.



Питання 2. В який час доби покупці ставлять найнижчі оцінки сервісу?

In [17]:
hourly_raiting = df.groupby('Hour')['Rating'].mean().sort_values()
print("Години з найнижчим рейтингом (топ-3):")
print(hourly_raiting.head(3))

Години з найнижчим рейтингом (топ-3):
Hour
19    6.716814
11    6.806667
16    6.859740
Name: Rating, dtype: float64


Для бізнесу: Виявили проблемні години (можливо, коли утворюються найбільші черги через брак персоналу в піковий час).

Питання 3. Якому способу оплати віддають перевагу постійні клієнти порівняно з звичайними покупцями і чи відрізняється їхній середній чек?

In [11]:
payment_by_customer = df.pivot_table(
    index='Customer type',
    columns='Payment',
    values='Invoice ID',
    aggfunc='count'
)

avg_check = df.groupby('Customer type')['Total'].mean()

print("Способи оплати за типом клієнта:")
print(payment_by_customer)
print("\nСередній чек (Total):")
print(avg_check)

Способи оплати за типом клієнта:
Payment        Cash  Credit card  Ewallet
Customer type                            
Member          168          172      161
Normal          176          139      184

Середній чек (Total):
Customer type
Member    327.791305
Normal    318.122856
Name: Total, dtype: float64


Для бізнесу: в залежності від політики компанії, можна проаналізувати, чи варто стимулювати клієнтів переходити на певний вид оплати(наприклад, робити знижки/кешбек за оплату карткою чи електрониим гаманцем або кеш), якщо це збільшує суму покупки.

Питання 4. Які категорії товарів найпопулярніші за кількістю серед чоловіків, а які - серед жінок?

In [13]:
gender_analysis = df.pivot_table(
    index='Product line',
    columns='Gender',
    values='Invoice ID',
    aggfunc='count'
)
print(gender_analysis)
print("\n")

Gender                  Female  Male
Product line                        
Electronic accessories      84    86
Fashion accessories         96    82
Food and beverages          90    84
Health and beauty           64    88
Home and lifestyle          79    81
Sports and travel           88    78




Для бізнесу: вже є певна картинка, яка дозволить ефективніше налаштувати таргетовану рекламу та планувати викладку товарів на полицях під конкретну аудиторію.

Питання 5. Товари яких категорій отримують найнижчі оцінки від чоловіків та жінок відповідно?

In [12]:
rating_by_gender = df.pivot_table(
    index='Product line',
    columns='Gender',
    values='Rating',
    aggfunc='mean'
).round(2)

print("Середній рейтинг товарів за статтю:")
print(rating_by_gender)

Середній рейтинг товарів за статтю:
Gender                  Female  Male
Product line                        
Electronic accessories    6.79  7.06
Fashion accessories       7.12  6.92
Food and beverages        7.20  7.02
Health and beauty         7.15  6.90
Home and lifestyle        6.76  6.91
Sports and travel         6.77  7.08


Для бізнесу: Аудит асортименту для проблемних категорій:

Home and lifestyle (найнижчий оціночний шар у жінок): Варто переглянути якість товарів, асортимент або співвідношення ціни та якості у цій категорії. Можливо, дизайн чи практичність товарів не відповідають очікуванням жіночої аудиторії.

Health and beauty (найнижчий оціночний шар у чоловіків): Оскільки для жінок ця категорія має один із найвищих рейтингів (7.15), а для чоловіків - найнижчий (6.90), є сенс розширити саме чоловічу лінійку догляду/косметики або змінити навігацію/викладку цих товарів у магазині, щоб чоловікам було простіше обирати потрібні позиції.

Персоналізація маркетингу та акцій:

Для чоловіків: Рекламні кампанії та спецпропозиції варто фокусувати навколо Sports and travel та Electronic accessories, оскільки ці категорії викликають найбільше позитивного відгуку.

Для жінок: Сильними магнітами для залучення є категорії Food and beverages та Health and beauty.